packages/block-editor/src/hooks/block-style-variation.js

```diff
@@ -59,7 +60,127 @@ function getVariationNameFromClass( className, registeredStyles = [] ) {
 	return null;
 }
 
-function useBlockSyleVariation( name, variation, clientId ) {
+// A helper component to apply a style override using the useStyleOverride hook.
+function OverrideStyles( { override } ) {
+	useStyleOverride( override );
+}
+
+/**
+ * This component is used to generate new block style variation overrides
+ * based on an incoming theme config. If a matching style is found in the config,
+ * a new override is created and returned. The overrides can be used in conjunction with
+ * useStyleOverride to apply the new styles to the editor. Its use is
+ * subject to change.
+ *
+ * @param {Object} props        Props.
+ * @param {Object} props.config A global styles object, containing settings and styles.
+ * @return {JSX.Element|undefined} An array of new block variation overrides.
+ */
+export function __unstableBlockStyleVariationOverridesWithConfig( { config } ) {
+	const { getBlockStyles, overrides } = useSelect(
+		( select ) => ( {
+			getBlockStyles: select( blocksStore ).getBlockStyles,
+			overrides: unlock( select( blockEditorStore ) ).getStyleOverrides(),
+		} ),
+		[]
+	);
+	const { getBlockName } = useSelect( blockEditorStore );
```

Comment:
So both `getBlockStyles` and `getBlockName` can't be moved into subscriptions? The catch here is that if any of these change, the overrides won't update.

Reply:
The blocks/block clientIds won't change between revisions.

We could map over the results of `getStyleOverrides()` in the `useSelect` callback and inject the block name into the overrides array items. 

So something like:

```js
	const { overrides, getBlockStyles } = useSelect(
		( select ) => {
	        const _overrides { getBlockName, getStyleOverrides } = unlock( select( blockEditorStore ) ) ;
			return {
				overrides: _overrides?.length ? 
					_overrides.map( ( [ id, override ] ) => 
						[ id, { ...override, blockName: getBlockName( override?.clientId ) } ] 
					) : [],
				getBlockStyles: select( blocksStore ).getBlockStyles,
			}
		},
		[]
	);

What do you reckon?

I think there's a flakey E2E test I'll have to take care of too as a follow up.

Reply:
I'll do the above in a follow up in trunk today or tomorrow first thing 👍🏻 with a view to merge back into wp/6.6

Reply:
Follow up https://github.com/WordPress/gutenberg/pull/62915

Reply:
Commented there. Also I think it's only really valuable if `getBlockStyles` can be changed too.